domehunter/dome_control.py

```diff
@@ -355,81 +386,176 @@ def calibrate_dome_encoder_counts(self, num_cal_rotations=2):
+
+        """
+        goingtoaz = False
+        if trigger_condition.__name__ == '_goto_az_condition':
+            goingtoaz = True
+        start = time.time()
+        while True:
+            time.sleep(0.1)
+            wait_time = time.time() - start
+            if trigger_condition():
+                logger.debug(f'Monitoring thread triggered by {trigger_condition.__name__}')
+                break
+            elif self._abort_event.is_set():
+                logger.debug(f'Monitoring thread triggered by abort event')
+                break
+            elif wait_time > self.wait_timeout:
+                logger.debug(f'Monitoring thread triggered by timeout [{self.wait_timeout}s]')
+                break
+            elif goingtoaz and self.testing is True:
+                # if testing simulate a tick for every cycle of while loop
+                self._simulate_ticks(num_ticks=1)
+        logger.debug('Monitor Thread: Stopping dome movement')
+        self._stop_moving()
+        # reset various dome state variables/events
+        self._move_event.clear()
+        self._calibrating = False
+        self.rotation_count = 0
+        return
+
+    def _goto_az_condition(self):
+        """Determines if current azimuth is within tolerance of target azimuth.
+
+        Returns
+        -------
+        bool
+            Returns true if self.dome_az is within tolerance of target azimuth.
+
+        """
+        delta_az = self.current_direction * (self.target_az - self.dome_az).wrap_at('180d')
+        logger.debug(f'Delta_az is {delta_az}')
+        return delta_az <= self.az_position_tolerance
+
+    def _find_home_condition(self):
+        """Return True if the dome is at home."""
+        # need a callable method for _thread_condition
+        return self._home_sensor.is_active
+
+    def _calibration_completed(self):
+        """Return True if both calibrating and move_event flags are True"""
+        cal_in_progress = self._calibrating and self._move_event.is_set()
+        return not cal_in_progress
+
+    def _calibrate(self):
+        """Calibration method to be called from within a non-blocking thread"""
+        start = time.time()
+        # if self.testing:
+        #     for i in range(self.num_cal_rotations):
+        #         pass
+        first_rotation = threading.Thread(target=self._simulate_rotation)
+        first_rotation.start()
+
+        while True:
+            logger.debug(f'Loop {time.time() - start}: rot_count [{self.rotation_count}], encoder_count [{self.encoder_count}]')
+            logger.debug(f'Simulated rotation underway {self._simulated_rotation_event.is_set()}')
+            if self.testing and not self._simulated_rotation_event.is_set():
+                logger.debug(f'Calibration monitor thread: Completion of simulated rotation detected, simulating next rotation.')
+                simulated_rotation = threading.Thread(target=self._simulate_rotation)
```

Comment:
Do you want another thread here? You are checking the `_simulated_rotation_event` in the conditional and never seem to use this thread either. I think you could just call `_simulate_rotation` directly (on line 536 below).

Reply:
Just realised this whole thing isn't really doing what is intended. The `cal_monitor` thread should be taking `_calibration_completed` as it's `trigger_condition` arg, and `_calibration_completed` should just be checking if we have completed the number of desired rotations. When `_calibrating` is True, `_set_at_home` will increment `rotation_count`.

The thread which I've called `calibrate` should just be the thread that starts simulated rotations one after the other and would only be needed in testing mode. It could terminate itself by also checking `_calibration_completed`. I think this is better as with real hardware, we can't just tell the dome to spin around n times. So have a simulator where we just call `_simulate_rotation` n times is not reflective of reality. We want that thread to start new rotations continually unless `_calibration_completed` returns True. 

so really that `calibrate` thread should be `simulate_calibrate` and be wrapped in an `if self.testing` clause.